### PR TITLE
fix CPP HeaderGenerator indentation

### PR DESCRIPTION
**Pull request by UplinkCoder**
This mainly fixed indentation and added namespace support

### CODE REVIEW

**Comment by thewilsonator** [comment]
Tests?

**Comment by UplinkCoder** [comment]
depends on https://github.com/dlang/dmd/pull/10975 to be merged first

**Comment by WalterBright** [comment]
I'm rejecting this because I really really want to emphasize that REFACTORING and BEHAVIOR changes must be in separate PRs. A few weeks ago I finally fixed a long standing regression introduced with a 350 line complex PR. I was totally daunted, but had to fix the regression.

I finally discovered that there was only one line of substantive change, the rest was all shuffling code about.

I've had enough of this. Refactorings, bug fixes, and enhancements go in separate PRs from now on.  Bug fixes also need a bugzilla issue. Also, this PR seems to mix in multiple different issues. Separate PRs for them, please.

**Comment by WalterBright** [comment]
I apologize as despite the subject message, this PR is not about refactoring.

**Comment by UplinkCoder** [comment]
@WalterBright you're right about this having to be broken up more.

I'll split this up once I am awake again.

File: src/dmd/dtoh.d [hunk 1]
```diff
@@ -564,12 +665,15 @@ public:
         if (adparent && fd.isAbstract())
             buf.writestring(" = 0");
         if (adparent && fd.isDisabled && global.params.cplusplus >= CppStdRevision.cpp11)
-            buf.printf(" = delete");
-        buf.printf(";\n");
+            buf.writestring(" = delete");
         if (adparent && fd.isDisabled && global.params.cplusplus < CppStdRevision.cpp11)
-            buf.printf("public:\n");
+            buf.printf("public:");
+        buf.printf(";");
```
Comment (thewilsonator) [inline src/dmd/dtoh.d]: This shouldn't change position. Otherwise it may produce `public:;`

**Comment by UplinkCoder** [comment]
@thewilsonator @Geod24 any comments? I feel this should be about ready.

**Comment by Geod24** [comment]
Conflict

File: src/dmd/root/outbuffer.d [hunk 2]
```diff
@@ -148,6 +148,12 @@ struct OutBuffer
         write(s);
     }
 
+    void writestringnl(const (char)[] s)
```
Comment (Geod24) [inline src/dmd/root/outbuffer.d]:
```suggestion
    void writestringln(const(char)[] s)
```

To match the existing functions

**Review by Geod24 — COMMENTED** [review]
It looks to me that the introduction of the `writestringnl` function should be first ? At the moment this is adding code then removing it.

**Comment by thewilsonator** [comment]
This needs a rebase.

**Comment by UplinkCoder** [comment]
Thanks done.

Am Mi., 1. Apr. 2020 um 14:14 Uhr schrieb Nicholas Wilson <
notifications@github.com>:

**Comment by thewilsonator** [comment]
Err, there are still conflicts.

**Review by thewilsonator — APPROVED** [review]
LGTM, @Geod24 ?

File: src/dmd/dtoh.d [hunk 3]
```diff
@@ -196,97 +230,125 @@ extern(C++) void genCppHdrFiles(ref Modules ms)
...
+        buf.writenl();
 
     buf.write(&done);
     buf.write(&decl);
 
     debug (Debug_DtoH)
     {
-        buf.writestring(`
-#if OFFSETS
+        buf.writestring(`#if OFFSETS
```
Comment (MoonlightSentinel) [inline src/dmd/dtoh.d:344]: Why not `hashIf`?

Comment (UplinkCoder) [reply src/dmd/dtoh.d:344]: just because I didn't want to touch this code block which is only in debug builds

File: src/dmd/root/outbuffer.d (hunk 2 above)
Comment (UplinkCoder) [reply src/dmd/root/outbuffer.d]: done

File: src/dmd/dtoh.d [hunk 4]
```diff
@@ -332,6 +396,8 @@ public:
         Other
     }
 
+    alias AST = ASTCodegen;
```
Comment (Geod24) [inline src/dmd/dtoh.d:399]: Why this ? Seems unrelated.

File: src/dmd/dtoh.d [hunk 5]
```diff
@@ -671,34 +744,32 @@ public:
...
-            buf.writestring(", ");
-            buf.writestring(vd.ident.toChars());
-            buf.printf(") == %d);\n", vd.offset);
-            buf = savex;
+            checkbuf.level++;
+            const pn = adparent.ident.toChars();
+            const vn = vd.ident.toChars();
+            const vo = vd.offset;
+            checkbuf.printf("assert(offsetof(%s, %s) == %d);",
+                                             pn, vn,    vo);
```
Comment (Geod24) [inline src/dmd/dtoh.d:767]: I see why you'd do that but it break auto-indentation for most people

**Review by Geod24 — APPROVED** [review]
I have two minor questions. The commit separation isn't great (it would be better to have the changes to `pragma(pack)` in a different PR). Otherwise very nice!

File: src/dmd/dtoh.d (hunk 5 above)
Comment (UplinkCoder) [reply src/dmd/dtoh.d:767]: I don't think the line will be touched.

File: src/dmd/dtoh.d (hunk 4 above)
Comment (UplinkCoder) [reply src/dmd/dtoh.d:399]: oh that. allows my editor to jup to the definitions of the AST structs.

Comment (UplinkCoder) [reply src/dmd/dtoh.d:399]: I can back this change out of you take offence at it

Comment (Geod24) [reply src/dmd/dtoh.d:399]: It's a functional change completely unrelated to indentation, so yes, please remove

File: src/dmd/dtoh.d (hunk 5 above)
Comment (Geod24) [reply src/dmd/dtoh.d:767]: Well it is at the moment, and the style you are using has no precedent in DMD.

File: test/compilable/dtoh_enum.d [hunk 6]
```diff
@@ -11,35 +11,35 @@ TEST_OUTPUT:
 #include <stdint.h>
 
 #if !defined(BEGIN_ENUM)
-# define BEGIN_ENUM(name, upper, lower) enum class name {
-# define ENUM_KEY(type, name, value, enumName, upper, lower, abbrev) name = value,
-# define END_ENUM(name, upper, lower) };
+    #define BEGIN_ENUM(name, upper, lower) enum class name {
+    #define ENUM_KEY(type, name, value, enumName, upper, lower, abbrev) name = value,
+    #define END_ENUM(name, upper, lower) };
```
Comment (TurkeyMan) [inline test/compilable/dtoh_enum.d:16]: According to the C spec, `#` must be at the start of the line. That's why people usually put the spaces after the `#`.

Comment (UplinkCoder) [reply test/compilable/dtoh_enum.d:16]: Oh ... that's unfortunate.

**Comment by TurkeyMan** [comment]
This is really helpful! Thanks @UplinkCoder !!

File: test/compilable/dtoh_enum.d (hunk 6 above)
Comment (UplinkCoder) [reply test/compilable/dtoh_enum.d:16]: I've never seen any modern c++ compiler rejecting it though, have you?

Comment (TurkeyMan) [reply test/compilable/dtoh_enum.d]: No, but these days it's more that C++ convention follows from that old requirement.
It's weird to indent the `#`, and people will think it looks amateur :P

**Comment by TurkeyMan** [comment]
Can you remove the fiddling with #define statements from this PR?
It's not related to namespace support, and conflates concerns. I'd really like this to be merged.
